drivers/w3s.go

```diff
@@ -0,0 +1,341 @@
+package drivers
+
+import (
+	"context"
+	"fmt"
+	bserv "github.com/ipfs/go-blockservice"
+	"github.com/ipfs/go-cid"
+	ds "github.com/ipfs/go-datastore"
+	dssync "github.com/ipfs/go-datastore/sync"
+	blockstore "github.com/ipfs/go-ipfs-blockstore"
+	format "github.com/ipfs/go-ipld-format"
+	"github.com/ipfs/go-merkledag"
+	"github.com/ipfs/go-unixfs"
+	"github.com/ipld/go-car"
+	"io"
+	"os"
+	"os/exec"
+	"regexp"
+	"strings"
+	"sync"
+	"time"
+)
+
+const w3SDefaultSaveTimeout = 2 * time.Minute
+
+var cidV1 = merkledag.V1CidPrefix()
+
+// This represents the main CAR directory structure organized by pubId.
+// Data for each pubId is removed after the CAR directory is published.
+var (
+	dataToPublish   = make(map[string]*rootCar)
+	dataToPublishMu sync.Mutex
+)
+
+type rootCar struct {
+	root    *merkledag.ProtoNode
+	dag     format.DAGService
+	carCids []string
+	mu      sync.Mutex
+}
+
+func newRootCar() *rootCar {
+	return &rootCar{
+		root: newDir(),
+		dag:  merkledag.NewDAGService(bserv.New(blockstore.NewBlockstore(dssync.MutexWrap(ds.NewMapDatastore())), nil)),
+	}
+}
+
+type W3sOS struct {
+	ucanKey   string
+	ucanProof string
+	dirPath   string
+	pubId     string
+}
+
+type W3sSession struct {
+	os *W3sOS
+}
+
+func NewW3sDriver(ucanKey, ucanProof, dirPath, pubId string) *W3sOS {
+	return &W3sOS{
+		ucanKey:   ucanKey,
+		ucanProof: ucanProof,
+		dirPath:   dirPath,
+		pubId:     pubId,
+	}
+}
+
+func (ostore *W3sOS) NewSession(filename string) OSSession {
+	if filename != "" {
+		panic("File names are not supported by W3S driver")
```

Comment:
I don't love having a `panic` inside a library, but can also see that changing the return to `(OSsession, error)` is going to result in a lot of pain for all the places we use this library. 

I'd vote for either doing that and refactoring wherever we call it as needed or returning a nil OSSession

Reply:
Yeah, agree. This is actually copied from other drivers.

I'll change it to `nil` in this PR and remove panics in all the code as a separate task, tracking here: https://github.com/livepeer/go-tools/issues/13

Reply:
Thanks!

Reply:
An alternative could be returning a "broken session" that returns errors for all function calls. The nil session should just cause a panic as well on code that ends up going this.

The basic thing I don't like about a panic or nil here is that it can be caused entirely by user input. The same code that works for an s3+https driver would explode is the user input is changed to w3s instead.

So I'd be fine with any handling of the filename here even if not super useful to avoid panicking on consumer code. Like returning the failed session, or maybe just prepending every file name in the CAR with the one provided for the session here.

Ok doing this as a follow-up or even just filling an issue to figure out a better error handling for the w3s session.

Reply:
Thanks for the comment @victorges 

Added your thoughts to the related GH Issue #13